osu.Game/Rulesets/Difficulty/Skills/Skill.cs

```diff
@@ -66,42 +84,43 @@ public void Process(DifficultyHitObject current)
         /// <summary>
         /// Saves the current peak strain level to the list of strain peaks, which will be used to calculate an overall difficulty.
         /// </summary>
-        public void SaveCurrentPeak()
+        private void saveCurrentPeak()
         {
-            if (Previous.Count > 0)
-                strainPeaks.Add(currentSectionPeak);
+            strainPeaks.Add(currentSectionPeak);
         }
 
         /// <summary>
         /// Sets the initial strain level for a new section.
         /// </summary>
         /// <param name="time">The beginning of the new section in milliseconds.</param>
-        public void StartNewSectionFrom(double time)
+        private void startNewSectionFrom(double time)
         {
             // The maximum strain of the new section is not zero by default, strain decays as usual regardless of section boundaries.
             // This means we need to capture the strain level at the beginning of the new section, and use that as the initial peak level.
-            if (Previous.Count > 0)
-                currentSectionPeak = GetPeakStrain(time);
+            currentSectionPeak = GetPeakStrain(time);
         }
 
         /// <summary>
         /// Retrieves the peak strain at a point in time.
         /// </summary>
         /// <param name="time">The time to retrieve the peak strain at.</param>
         /// <returns>The peak strain.</returns>
-        protected virtual double GetPeakStrain(double time) => CurrentStrain * strainDecay(time - Previous[0].BaseObject.StartTime);
+        protected virtual double GetPeakStrain(double time) => CurrentStrain * strainDecay(time - Previous[0].StartTime);
 
         /// <summary>
-        /// Returns the calculated difficulty value representing all processed <see cref="DifficultyHitObject"/>s.
+        /// Returns the calculated difficulty value representing all <see cref="DifficultyHitObject"/>s that have been processed up to this point.
         /// </summary>
         public double DifficultyValue()
         {
             double difficulty = 0;
             double weight = 1;
 
+            var finalStrainPeaks = strainPeaks.ToList(); // Copy list so we don't create side effects
+            finalStrainPeaks.Add(currentSectionPeak); // Add final peak since it never gets saved
```

Comment:
I'm not sure I understand this comment. The final strain peak is already added at the end:
https://github.com/ppy/osu/blob/e8f08b09dcc302652a66b5476fe9c1fd7ec8e43a/osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs#L96-L98

If I'm missing something here then this is a huge bug, because I've already had to address this in osu-stable too to match lazer (with a completely different diffcalc structure).

Reply:
Yes, and current `master` is correct. What I'm getting at is that taiko diffcalc at one point queries the skills' `StrainPeak`s, and due to this code not saving the last peak to that protected list as was done previously (to "avoid side effects"), it changes taiko results.

Reply:
Ah I see, nice catch.
Fixed in 3743937774fb21887ee2472555274a5784d44048

I've opted to expose `CurrentStrainPeak` as a property as opposed to modifying `StrainPeaks` to add the final incomplete section's peak for 2 reason.

1. A dynamic `StrainPeaks` property could be a little confusing if other expect it to directly reflect the internal `strainPeaks` field.
2. It seems like a bad idea to copy an entire list for the sake of a dynamic property when efficiency is a key objective.

Reply:
You could use `Enumerable.Append` to avoid a copy.

Not super sure it's reasonable to make consumers remember to include the last peak themselves, to be honest.

Reply:
Ooh never knew about that method. Pretty neat.
Yea, I'm happy to use that instead. I'll change the name of the property though to make it more obvious that its more nuanced than a mere getter.

Fixed in 01bde91a5b2b1c9aea51c434fe1426042e2b8de9